### ISSUE TITLE
(Android) Crash when changing core's options

### ISSUE THREAD

**Issue by Zlika** [libretro/RetroArch#7130]
When using the core theodore, RetroArch 1.7.3 (Android version) crashes each time I try to change one of the core's option.
There is no problem when using the Linux version of RetroArch. Also, the crash occurs on two phones I tested (one under Android 8 and another under Android 7), but not on a Tablet running Android 7.
It may be a bug on the core, however I cannot find any problem on it, even when using Valgrind on Linux to detect potential memory read/write problems.
I have a detailed report on the bug (cf. below). It is a null pointer dereference that occurs in the menu_input_post_iterate function.

Thanks for your help.

```Build fingerprint: 'HUAWEI/PRA-LX1/HWPRA-H:8.0.0/HUAWEIPRA-LX1/370(C432):user/release-keys'
Revision: '0'
ABI: 'arm'
pid: 7090, tid: 7111, name: Thread-5  >>> com.retroarch <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x4
Cause: null pointer dereference
    r0 00000000  r1 00000000  r2 00005000  r3 00000001
    r4 c5967000  r5 de0eda40  r6 00000168  r7 c705c6c8
    r8 00000001  r9 00000092  sl 00000005  fp c6deb5f8
    ip d87b08c0  sp c6deb5f0  lr c7089b50  pc c705c6f8  cpsr a00d0010
    d0  72635f7343b40000  d1  706f5f6343120000
    d2  40d046aa43dbdc48  d3  000000004312dc48
    d4  0000000000000000  d5  0000000000000000
    d6  0000000000000000  d7  0000000000000000
    d8  0000000000000000  d9  0000000000000000
    d10 0000000000000000  d11 0000000000000000
    d12 0000000000000000  d13 0000000000000000
    d14 0000000000000000  d15 0000000000000000
    d16 000000020000003c  d17 0000000100000001
    d18 0000000003980a30  d19 0000000004108fb0
    d20 bf67a000bf67a000  d21 bf67a000bf67a000
    d22 bf67a000bf67a000  d23 bf67a000bf67a000
    d24 3d36d927bc35df80  d25 3f648e19be0917c4
    d26 bdc376d93e8cd794  d27 bb8cbe6a3d0ccc82
    d28 0000000000000000  d29 0000000000000000
    d30 00000000735a11e4  d31 000008d0000008c4
    scr 8b000011

backtrace:
    #00 pc 002706f8  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so
    #01 pc 0029db4c  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so
    #02 pc 002493f8  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so (menu_input_post_iterate+3868)
    #03 pc 0028aae8  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so (generic_menu_iterate+2208)
    #04 pc 00245c88  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so (menu_driver_iterate+272)
    #05 pc 001db0a8  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so (runloop_iterate+1564)
    #06 pc 001da918  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so (rarch_main+296)
    #07 pc 002fc844  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so
    #08 pc 002201e4  /data/app/com.retroarch-oW8E8dz2ksH6nxG1X7-4rw==/lib/arm/libretroarch-activity.so
    #09 pc 000486cf  /system/lib/libc.so (_ZL15__pthread_startPv+22)
    #10 pc 0001b02f  /system/lib/libc.so (__start_thread+32)

- RetroArch: 1.7.3

### Environment information

- OS: Android

**Comment by bparker06** [comment]
The backtrace shows that it's crashing inside RA itself and not the core, so you would need to build a debug version of RA for Android and get a new backtrace with that.

**Comment by Zlika** [comment]
Hi @bparker06. Thanks for your answer.

I know that the crash occurs inside RA, however as this problem does not seem to exist with other cores, I think it may be linked to the way options are declared in this core:

```
static const struct retro_variable prefs[] = {
    { PACKAGE_NAME"_rom", "Thomson flavor; TO8|TO8D|TO9+" },
    { PACKAGE_NAME"_floppy_write_protect", "Floppy write protection; enabled|disabled" },
    { PACKAGE_NAME"_tape_write_protect", "Tape write protection; enabled|disabled" },
    { PACKAGE_NAME"_printer_emulation", "Dump printer data to file; disabled|enabled" },
    { NULL, NULL }
};
```

The only thing special I see is that there is a '+' in one of the option strings: maybe this char is not correctly parsed by RA, or it is a reserved character that should not be used here?
The crash occurs systematically as soon as I touch the screen to change one of the options.

I'll try to compile and install a debug version of RA, but as I've never done that before I have to take some time to read some docs about it.

As there are 2 unamed functions below menu_input_post_iterate in the backstrace, I suppose they are static functions so my guess is that the crash maybe happens in either menu_input_mouse_frame() or menu_input_pointer_post_iterate().
As the crash happens on a phone with Android 7 but not on a tablet with Android 7, do you know if there is some code here that could be different between a tablet and a phone?

Regards

**Comment by Zlika** [comment]
Ok, I was able to break into core_option_manager_next().
```void core_option_manager_next(core_option_manager_t *opt, size_t idx)
{
   struct core_option *option = NULL;

   if (!opt)
      return;

   option        = (struct core_option*)&opt->opts[idx];

   option->index = (option->index + 1) % option->vals->size;
   opt->updated  = true;
}
```
When the bug occurs, idx = 4294922240 which is not a realistic option number (the core only has 4 options).

Previous function in the stack is rarch_ctl():
```case RARCH_CTL_CORE_OPTION_NEXT:
         {
            unsigned *idx = (unsigned*)data;
            if (!idx)
               return false;
            core_option_manager_next(runloop_core_options, *idx);
         }
```
idx = 0xd5fcfab4 and *idx = 4294922240.
Previous function in the stack is core_setting_right():
```int core_setting_right(unsigned type, const char *label,
      bool wraparound)
{
   unsigned idx     = type - MENU_SETTINGS_CORE_OPTION_START;

   rarch_ctl(RARCH_CTL_CORE_OPTION_NEXT, &idx);

   return 0;
}
```
Here we have type = 20480, label = "game_specific_options_create", wraparound = true.
Before that we have function action_select_core_setting(), then menu_entry_action():
```case MENU_ACTION_SELECT:
         if (cbs && cbs->action_select)
            ret = cbs->action_select(entry->path,
                  entry->label, entry->type, i);
```
Here, entry->path = "Create game-options file", entry->label = "game_specific_options_create", entry->type = 20480 and i = 3.

The code is not very clear to me here. Could someone explain to me why the option number in core_option_manager_next is out of bound?
Thank you.

**Comment by bparker06** [comment]
The backtrace still does not show line numbers, I think something may be wrong with your debug build. Try rebuilding with this patch: https://github.com/libretro/RetroArch/blob/master/retroarch-android-debug.diff

Regarding the wrong `idx` value, you need to figure out at what point it is changed to that number. You may have to work your way backwards through the backtrace to figure out how that number is calculated before it's passed to the `core_option_manager_next()` function.

**Comment by Zlika** [comment]
**I was able to reproduce the exact same bug with other cores: ChaiLove, Mr. Boom...** So it seems to be a "real" bug.

The null pointer dereference occurs in core_option_manager.c at line 365 (```option->index = (option->index + 1) % option->vals->size;```). It is cause because at the line just before, the index of the core option in the opt->opts array is much too big (idx=4294922240, where the cores only have a few options). This value is computed in menu_cbs_right.c at line 495 (```unsigned idx     = type - MENU_SETTINGS_CORE_OPTION_START;```).
Here type = 20480 (i.e. the value of MENU_SETTINGS_CORE_OPTION_CREATE as defined in menu_driver.h) and MENU_SETTINGS_CORE_OPTION_START = 65536 (as defined in menu_driver.h), so the difference is negative and casted into a (big) unsigned int.

When testing with a core that does not trigger this bug (e.g. Nestopia), the value of type is 65536 (i.e. the value of MENU_ENUM_LABEL_CORE_OPTION_ENTRY as defined in msg_hash.h) and then idx starts at 0.

**Comment by Zlika** [comment]
I think I've found the problem.

In menu_input_pointer_post_iterate(), menu_input->pointer.ptr is used to know the menu item selected. However, this value is wrong because it is an old value related to the previous menu item selected. If this "old" value is still a "valid" value for the new menu, then the bug is unnoticed. However, if this value is not a valid value, then the application crashes. When we enter the core's option menu, we've just selected the 10th menu item ("Options"). Then, if the core has less than 10 options, this old "10" value is not a valid value for the new option's menu and the application crashes. That's why we can reproduce this bug with some cores but not all.

I traced the root cause in runloop_check_state (retroarch.c):
* at line 2731, pointer.x and pointer.y will be updated (but not pointer.ptr)
* at line 2775, the event generated (menu_driver_ctl(RARCH_MENU_CTL_POINTER_DOWN, &point) in menu_input_pointer_post_iterate()) uses pointer.ptr, but as it was not updated yet it's still the old value from the previous selection!
* at line 3387, pointer.ptr is computed based on pointer.y, but too late!

However I still don't know how to fix the bug without breaking the app.
